Source/WebKit/UIProcess/WebProcessCache.cpp

```diff
@@ -207,7 +237,55 @@ RefPtr<WebProcessProxy> WebProcessCache::takeProcess(const WebCore::Site& site,
     }
 
     Ref process = m_processesPerSite.take(it)->takeProcess();
-    WEBPROCESSCACHE_RELEASE_LOG("takeProcess: Taking process from WebProcess cache (size=%u, capacity=%u, processWasTerminated=%d) %" PRIVATE_LOG_STRING, process->processID(), size(), capacity(), process->wasTerminated(), site.toString().utf8().data());
+    WEBPROCESSCACHE_RELEASE_LOG("takeProcess: Taking process from WebProcess cache (size=%u, capacity=%u, processWasTerminated=%d) %" SENSITIVE_LOG_STRING, process->processID(), size(), capacity(), process->wasTerminated(), site.toString().utf8().data());
+
+    ASSERT(!process->pageCount());
+    ASSERT(!process->provisionalPageCount());
+    ASSERT(!process->suspendedPageCount());
+
+    if (process->wasTerminated()) {
+        WEBPROCESSCACHE_RELEASE_LOG("takeProcess: cannot take process, was terminated", process->processID());
+        return nullptr;
+    }
+
+    return process;
+}
+
+RefPtr<WebProcessProxy> WebProcessCache::takeSharedProcess(const WebCore::Site& mainFrameSite, const HashSet<WebCore::RegistrableDomain>& isolatedDomains, WebsiteDataStore& dataStore, WebProcessProxy::LockdownMode lockdownMode, WebProcessProxy::EnhancedSecurity enhancedSecurity, const API::PageConfiguration& pageConfiguration)
+{
+    auto it = m_sharedProcessesPerSite.find(mainFrameSite);
+    if (it == m_sharedProcessesPerSite.end()) {
+        WEBPROCESSCACHE_RELEASE_LOG("takeSharedProcess: did not find %" SENSITIVE_LOG_STRING, 0, mainFrameSite.toString().utf8().data());
+        return nullptr;
+    }
+
+    if (it->value->process().websiteDataStore() != &dataStore) {
+        WEBPROCESSCACHE_RELEASE_LOG("takeSharedProcess: cannot take process, datastore not identical", it->value->process().processID());
+        return nullptr;
+    }
+
+    if (it->value->process().lockdownMode() != lockdownMode) {
+        WEBPROCESSCACHE_RELEASE_LOG("takeSharedProcess: cannot take process, lockdown mode not identical", it->value->process().processID());
+        return nullptr;
+    }
+
+    if (it->value->process().enhancedSecurity() != enhancedSecurity) {
+        WEBPROCESSCACHE_RELEASE_LOG("takeSharedProcess: cannot take process, enhanced security not identical", it->value->process().processID());
+        return nullptr;
+    }
+
+    if (!it->value->process().sharedProcessDomains().intersectionWith(isolatedDomains).isEmpty()) {
+        WEBPROCESSCACHE_RELEASE_LOG("takeSharedProcess: cannot take process, contains a domain that needs to be isolated", it->value->process().processID());
+        return nullptr;
+    }
```

Comment:
It seems like at this point, instead of just returning `nullptr`, you would also want to remove the process from the cache, since it seems like there's no way we could ever use that cached process in the future.

Reply:
Ok, I'll move up to `WebProcessPool::processForSite` since that's where we create a new process if there is no match in the cache.